File: DifferentiablePointRender/GaussianSplatting/src/compute_cov3d.cpp

/**
 * @file compute_cov3d.cpp
 * @author Jian Gao
 * @brief Wrapper function of 3D covariance matrices coputation.
 * @version 0.1
 * @date 2024-01-18
 * 
 * @copyright Copyright (c) 2024
 * 
 */

#include <utils.h>
#include <compute_cov3d.h>
#include <compute_cov3d_kernel.h>

torch::Tensor
computeCov3DForward(
  const torch::Tensor& scales,
  const torch::Tensor& uquats,
  const torch::Tensor& visibility_status
){

    CHECK_INPUT(scales);
    CHECK_INPUT(uquats);
    CHECK_INPUT(visibility_status);

    const int P = scales.size(0);
    auto float_opts = scales.options().dtype(torch::kFloat32);
    torch::Tensor cov3Ds = torch::zeros({P, 6}, float_opts);
    if(P != 0)
    {
        computeCov3DForwardCUDA(
            P,
            scales.contiguous().data_ptr<float>(),
            uquats.contiguous().data_ptr<float>(),
            visibility_status.contiguous().data_ptr<bool>(),
            cov3Ds.data_ptr<float>());
    }

    return cov3Ds;
}



std::tuple<torch::Tensor, torch::Tensor>
computeCov3DBackward(
    const torch::Tensor& scales,
	const torch::Tensor& uquats,
	const torch::Tensor& visibility_status,
	const torch::Tensor& dL_dcov3Ds
){

    CHECK_INPUT(scales);
    CHECK_INPUT(uquats);
    CHECK_INPUT(visibility_status);
    CHECK_INPUT(dL_dcov3Ds);

    const int P = scales.size(0);
    auto float_opts = scales.options().dtype(torch::kFloat32);
    torch::Tensor dL_dscales = torch::zeros({P, 3}, float_opts);
    torch::Tensor dL_duquats = torch::zeros({P, 4}, float_opts);
    
    if(P != 0)
    {
        computeCov3DBackwardCUDA(
                P,
                scales.contiguous().data_ptr<float>(),
                uquats.contiguous().data_ptr<float>(),
                visibility_status.contiguous().data_ptr<bool>(),
                dL_dcov3Ds.contiguous().data_ptr<float>(),
                dL_dscales.data_ptr<float>(),
                dL_duquats.data_ptr<float>()
        );
    }

    return std::make_tuple(dL_dscales, dL_duquats);
}

File: DifferentiablePointRender/GaussianSplatting/compute_cov3d.py

import torch
from torch import Tensor
from jaxtyping import Float
import DifferentiablePointRender.GaussianSplatting._C as _C


def compute_cov3d(
    scales: Float[Tensor, "*batch 3"],
    uquats: Float[Tensor, "*batch 4"],
    visibility_status: Float[Tensor, ""] = None,
)->Tensor:
    """Compute the 3D covariance matrix.

    Args:
        scales (Tensor): _description_
        uquats (Tensor): _description_
        visibility_status (Tensor, optional): _description_. Defaults to None.

    Returns:
        - **cov3d** (Tensor): 3D covariance vector, upper right part of the covariance matrix.
    """
    if visibility_status is None:
            visibility_status = torch.ones_like(scales[:, 0], dtype=torch.bool)
    
    return _ComputeCov3D.apply(
        scales,
        uquats,
        visibility_status
    )


class _ComputeCov3D(torch.autograd.Function):
    @staticmethod
    def forward(
        ctx, 
        scales, 
        uquats, 
        visibility_status):
        
        cov3Ds = _C.compute_cov3d_forward(
            scales, 
            uquats, 
            visibility_status
        )
        
        # save variables for backward
        ctx.save_for_backward(
            scales, 
            uquats, 
            visibility_status)
        
        return cov3Ds

    @staticmethod
    def backward(ctx, dL_dcov3Ds):
        # get saved variables from forward
        scales, uquats, visibility_status = ctx.saved_tensors
        
        (
            dL_dscales, 
            dL_duquats
        ) = _C.compute_cov3d_backward(
            scales, 
            uquats, 
            visibility_status, 
            dL_dcov3Ds
        )

        grads = (
            # loss gradient w.r.t scales
            dL_dscales,
            # loss gradient w.r.t uquats
            dL_duquats,
            # loss gradient w.r.t visibility_status
            None,
        )

        return grads
